Rollbar/RollbarLogger.cs

```diff
@@ -502,142 +509,218 @@ void IDisposable.Dispose()
+                    new PayloadBundle(this.Config, rollbarPackage, level, custom, timeoutAt, signal);
+            }
+            else
             {
-                var data = RollbarUtility.PackageAsPayloadData(utcTimestamp, this.Config, level, dataObject, custom);
-                var payload = new Payload(this._config.AccessToken, data, timeoutAt, signal);
-                DoSend(payload);
+                payloadBundle =
+                    new PayloadBundle(this.Config, dataObject, level, custom, timeoutAt, signal);
             }
-        }
 
-        private void DoSend(Payload payload)
-        {
-            //lock (this._syncRoot)
+            if (payloadBundle == null)
             {
-                // here is the last chance to decide if we need to actually send this payload
-                // based on the current config settings:
-                if (string.IsNullOrWhiteSpace(this._config.AccessToken)
-                    || this._config.Enabled == false
-                    || (this._config.LogLevel.HasValue && payload.Data.Level < this._config.LogLevel.Value)
-                    )
-                {
-                    return;
-                }
-
-                if (TelemetryCollector.Instance.Config.TelemetryEnabled)
-                {
-                    payload.Data.Body.Telemetry =
-                        TelemetryCollector.Instance.GetQueueContent();
-                }
-
-                if (this._config.Server != null)
-                {
-                    payload.Data.Server = this._config.Server;
-                }
-
-                try
-                {
-                    if (this._config.CheckIgnore != null
-                        && this._config.CheckIgnore.Invoke(payload)
-                        )
-                    {
-                        return;
-                    }
-                }
-                catch (System.Exception ex)
-                {
-                    OnRollbarEvent(new InternalErrorEventArgs(this, payload, ex, "While  check-ignoring a payload..."));
-                }
-
-                try
-                {
-                    this._config.Transform?.Invoke(payload);
-                }
-                catch (System.Exception ex)
-                {
-                    OnRollbarEvent(new InternalErrorEventArgs(this, payload, ex, "While  transforming a payload..."));
-                }
-
-                try
-                {
-                    this._config.Truncate?.Invoke(payload);
-                }
-                catch (System.Exception ex)
-                {
-                    OnRollbarEvent(new InternalErrorEventArgs(this, payload, ex, "While  truncating a payload..."));
-                }
-
-                this._payloadQueue.Enqueue(payload);
+                //TODO: we may want to report that there is some problem with packaging...
+                return this;
+            }
 
-                return;
+            lock (this._syncRoot)
```

Comment:
What is the point of this lock when inside `Enqueue` the first thing we do is lock?

Reply:
oh, yes, that is the queue is thread safe on its own. i'll be cleaning it soon...